blockchain/blockchain.go

```diff
@@ -881,64 +885,27 @@ func (b *Blockchain) executeBlockTransactions(block *types.Block) (*BlockResult,
 func (b *Blockchain) WriteFullBlock(fblock *types.FullBlock, source string) error {
 	block := fblock.Block
 
-	b.writeLock.Lock()
-	defer b.writeLock.Unlock()
-
-	if block.Number() <= b.Header().Number {
-		b.logger.Info("block already inserted", "block", block.Number(), "source", source)
-
-		return nil
-	}
-
-	header := block.Header
-
-	if err := b.writeBody(block); err != nil {
-		return err
-	}
-
-	// Write the header to the chain
-	evnt := &Event{Source: source}
-	if err := b.writeHeaderImpl(evnt, header); err != nil {
-		return err
-	}
-
-	// write the receipts, do it only after the header has been written.
-	// Otherwise, a client might ask for a header once the receipt is valid,
-	// but before it is written into the storage
-	if err := b.db.WriteReceipts(block.Hash(), fblock.Receipts); err != nil {
-		return err
-	}
-
-	// update snapshot
-	if err := b.consensus.ProcessHeaders([]*types.Header{header}); err != nil {
-		return err
-	}
-
-	b.dispatchEvent(evnt)
-
-	// Update the average gas price
-	b.updateGasPriceAvgWithBlock(block)
-
-	logArgs := []interface{}{
-		"number", header.Number,
-		"txs", len(block.Transactions),
-		"hash", header.Hash,
-		"parent", header.ParentHash,
-	}
-
-	if prevHeader, ok := b.GetHeaderByNumber(header.Number - 1); ok {
-		diff := header.Timestamp - prevHeader.Timestamp
-		logArgs = append(logArgs, "generation_time_in_seconds", diff)
-	}
-
-	b.logger.Info("new block", logArgs...)
-
-	return nil
+	return b.writeBlockInternal(block, source,
```

Comment:
What is the benefit of having this with the callback? Having fewer lines of code and merging both functions? As I mentioned, we are transitioning from ibft (in prod) to polybft (in dev).

In the near future, `WriteBlock` is going to go away, it is only here right now because `ibft` relies on that (and its internal cache to some extend). This callback method overcomplicates the logic to simplify a code that will be legacy pretty soon.

Were these changes required for these two functions at all?

Reply:
I see this was part of this commit 62166e6877f5a6a6f4a780d7047511caa530b8df. Could we revert it?

Reply:
If it is going to be legacy soon or not, that is a separate topic. However having ~90% same code in two functions isn't good coding practice at all (hope we agree on that). That's why I opted to change it anyway, because I don't see it makes code any harder for maintenance and following.

Anyhow, your call, we can drop that commit easily if you don't like it :slightly_smiling_face:

Reply:
Ok hopefully we are going to deprecate `WriteBlock` any time soon, because my concern is that it is going to be there forever and it is not nice having two (almost) same implementations :slightly_smiling_face:

Reply:
Reverted: https://github.com/0xPolygon/polygon-edge/pull/1110/commits/c7d6bdb11957cae8d473049af496a17dc223ca71